Current MAME core on Lakka Switch or Any Platform.

https://github.com/libretro/mame
Here is the Project Link. Current MAME core is on Linux and android. so It might be easy to Port Lakka core. What do you Think about this?
and Nothing to Working Game in 2015 core. I Tried CV1K games and Black Screen.

I tried to build it for Switch - it uses the target toolchain to build a binary (`genie`), that is needed on the host and than wants to use that binary to build:
```
3rdparty/genie/bin/linux/genie --distro=debian-stable --OPTIMIZE=3 --NOWERROR='1' --target='mame' --subtarget='arcade' --build-dir='build' --PYTHON_EXECUTABLE='python2' --osd='retro' --targetos='linux' --PLATFORM='aarch64' --gcc=linux-gcc --gcc_version=9.4.0 gmake
3rdparty/genie/bin/linux/genie: 1: Syntax error: "(" unexpected
make: *** [makefile:1438: build/projects/retro/mamearcade/gmake-linux/Makefile] Error 2
make: *** Waiting for unfinished jobs....
```

```
$ file build.Lakka-Switch.aarch64/mame-7b6d18f/3rdparty/genie/bin/linux/genie
build.Lakka-Switch.aarch64/mame-7b6d18f/3rdparty/genie/bin/linux/genie: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, stripped
```
The host is x86_64, but the binary is for aarch64... we have to first build genie with host-toolchain and then continue the build for target. I personally do not know how to proceed from here, but this is my "finding".

--- a/distributions/Lakka/options
+++ b/distributions/Lakka/options
@@ -248,6 +248,7 @@
                   lowres_nx \
                   lr_moonlight \
                   lutro \
+                  mame \
                   mame2000 \
                   mame2003_plus \
                   mame2010 \
@@ -343,6 +344,7 @@
                                higan_sfc_balanced \
                                kronos \
                                lr_moonlight \
+                               mame \
                                mame2003_plus \
                                mame2010 \
                                mame2015 \

--- a/packages/lakka/libretro_cores/mame/package.mk
+++ b/packages/lakka/libretro_cores/mame/package.mk
@@ -3,28 +3,42 @@ PKG_VERSION="1f6e6075e9668f55d93c5ef4fa3b032c00c0b305"
 PKG_LICENSE="MAME"
 PKG_SITE="https://github.com/libretro/mame"
 PKG_URL="${PKG_SITE}.git"
-PKG_DEPENDS_TARGET="toolchain"
+PKG_DEPENDS_TARGET="toolchain expat zlib flac sqlite"
 PKG_LONGDESC="MAME - Multiple Arcade Machine Emulator"
-#PKG_BUILD_FLAGS="-gold -lto"
-PKG_TOOLCHAIN="manual"
+PKG_TOOLCHAIN="make"
+
+PKG_MAKE_OPTS_TARGET="-f Makefile.libretro REGENIE=1 VERBOSE=1 NOWERROR=1 OPENMP=0 CROSS_BUILD=1 TOOLS=0 RETRO=1 PYTHON_EXECUTABLE=python3 CONFIG=libretro LIBRETRO_OS=unix TARGET=mame OSD=retro USE_SYSTEM_LIB_EXPAT=1 USE_SYSTEM_LIB_ZLIB=1 USE_SYSTEM_LIB_FLAC=1 USE_SYSTEM_LIB_SQLITE3=1 LIBRETRO_CPU= ARCH= PROJECT="
+
+case ${ARCH} in
+  x86_64)
+    PKG_MAKE_OPTS_TARGET+=" NOASM=0 PTR64=1 PLATFORM=x86_64"
+    ;;
+  i386)
+    PKG_MAKE_OPTS_TARGET+=" NOASM=0 PTR64=0 PLATFORM=x86"
+    ;;
+  aarch64)
+    PKG_MAKE_OPTS_TARGET+=" NOASM=0 PTR64=0 PLATFORM=arm64"
+    ;;
+  arm)
+    PKG_MAKE_OPTS_TARGET+=" NOASM=1 PTR64=0 PLATFORM=arm"
+    ;;
+esac
+
+pre_make_target() {
+  PKG_MAKE_OPTS_TARGET+=" OVERRIDE_CC=${CC} OVERRIDE_CXX=${CXX} OVERRIDE_LD=${LD}"
+  sed -i scripts/genie.lua \
+      -e 's|-static-libstdc++||g'
+}
 
 make_target() {
-  LCPU=${ARCH}
-  PTR64=0
-  NOASM=0
-
-  if [ "${ARCH}" = "arm" ]; then
-    NOASM=1
-  elif [ "${ARCH}" = "i386" ]; then
-    LCPU=x86
-  elif [ "${ARCH}" = "x86_64" ]; then
-    PTR64=1
-  fi
-
-  make REGENIE=1 VERBOSE=1 NOWERROR=1 PYTHON_EXECUTABLE=python2 CONFIG=libretro LIBRETRO_OS="unix" ARCH="" PROJECT="" LIBRETRO_CPU="${LCPU}" DISTRO="debian-stable" CC="${CC}" CXX="${CXX}" LD="${LD}" CROSS_BUILD="" PTR64="${PTR64}" TARGET="mame" SUBTARGET="arcade" PLATFORM=${LCPU} RETRO=1 OSD="retro"
+  unset DISTRO
+  make ${PKG_MAKE_OPTS_TARGET}
 }
 
 makeinstall_target() {
   mkdir -p ${INSTALL}/usr/lib/libretro
-    cp -v *_libretro.so ${INSTALL}/usr/lib/libretro/mame_libretro.so
+    cp -v mame_libretro.so ${INSTALL}/usr/lib/libretro/
+
+  mkdir -p ${INSTALL}/usr/share/retroarch-system/mame
+    cp -vr artwork samples ${INSTALL}/usr/share/retroarch-system/mame
 }

--- a/packages/lakka/libretro_cores/mame/patches/mame-crosscompile.patch
+++ b/packages/lakka/libretro_cores/mame/patches/mame-crosscompile.patch
@@ -1,0 +1,54 @@
+--- a/3rdparty/genie/build/gmake.linux/genie.make	2018-08-11 00:17:44.000000000 +0200
++++ b/3rdparty/genie/build/gmake.linux/genie.make	2019-04-03 11:57:27.671223016 +0200
+@@ -28,8 +28,8 @@
+   RM    = $(SILENT) del /F "$(subst /,\\,$(1))" 2> nul || exit 0
+ endif
+ 
+-CC  = gcc
+-CXX = g++
++CC  = $(HOST_CC)
++CXX = $(HOST_CXX)
+ AR  = ar
+ 
+ ifndef RESCOMP
+@@ -50,12 +50,12 @@
+   INCLUDES           += -I"../../src/host/lua-5.3.0/src"
+   ALL_CPPFLAGS       += $(CPPFLAGS) -MMD -MP -MP $(DEFINES) $(INCLUDES)
+   ALL_ASMFLAGS       += $(ASMFLAGS) $(CFLAGS) $(ALL_CPPFLAGS) $(ARCH) -Wall -Wextra -Os $(MPARAM) -Wno-implicit-fallthrough
+-  ALL_CFLAGS         += $(CFLAGS) $(ALL_CPPFLAGS) $(ARCH) -Wall -Wextra -Os $(MPARAM) -Wno-implicit-fallthrough
+-  ALL_CXXFLAGS       += $(CXXFLAGS) $(CFLAGS) $(ALL_CPPFLAGS) $(ARCH) -Wall -Wextra -Os $(MPARAM) -Wno-implicit-fallthrough
+-  ALL_OBJCFLAGS      += $(CFLAGS) $(ALL_CPPFLAGS) $(ARCH) -Wall -Wextra -Os $(MPARAM) -Wno-implicit-fallthrough
+-  ALL_OBJCPPFLAGS    += $(CXXFLAGS) $(CFLAGS) $(ALL_CPPFLAGS) $(ARCH) -Wall -Wextra -Os $(MPARAM) -Wno-implicit-fallthrough
++  ALL_CFLAGS         += $(ALL_CPPFLAGS) $(ARCH) -Wall -Wextra -Os $(MPARAM) -Wno-implicit-fallthrough
++  ALL_CXXFLAGS       += $(ALL_CPPFLAGS) $(ARCH) -Wall -Wextra -Os $(MPARAM) -Wno-implicit-fallthrough
++  ALL_OBJCFLAGS      += $(ALL_CPPFLAGS) $(ARCH) -Wall -Wextra -Os $(MPARAM) -Wno-implicit-fallthrough
++  ALL_OBJCPPFLAGS    += $(ALL_CPPFLAGS) $(ARCH) -Wall -Wextra -Os $(MPARAM) -Wno-implicit-fallthrough
+   ALL_RESFLAGS       += $(RESFLAGS) $(DEFINES) $(INCLUDES)
+-  ALL_LDFLAGS        += $(LDFLAGS) -L"." -s -rdynamic
++  ALL_LDFLAGS        += -L"." -s -rdynamic
+   LIBDEPS            +=
+   LDDEPS             +=
+   LDRESP              =
+--- a/3rdparty/genie/src/host/scripts.c	2019-05-21 22:41:14.207745861 +0200
++++ b/3rdparty/genie/src/host/scripts.c	2019-05-21 22:42:43.159893481 +0200
+@@ -147,7 +147,7 @@
+ 
+ 	/* tools/gcc.lua */
+ 	"premake.gcc = { }\npremake.gcc.cc     = \"gcc\"\npremake.gcc.cxx    = \"g++\"\npremake.gcc.ar     = \"ar\"\npremake.gcc.rc     = \"windres\"\npremake.gcc.llvm   = false\nlocal cflags =\n{\nEnableSSE        = \"-msse\",\nEnableSSE2       = \"-msse2\",\nEnableAVX        = \"-mavx\",\nEnableAVX2       = \"-mavx2\",\nPedanticWarnings = \"-Wall -Wextra -pedantic\",\nExtraWarnings    = \"-Wall -Wextra\",\nFatalWarnings    = \"-Werror\",\nFloatFast        = \"-ffast-math\",\nFloatStrict      = \"-ffloat-store\",\nNoFramePointer   = \"-fomit-frame-pointer\",\nOptimize         = \"-O2\",\nOptimizeSize     = \"-Os\",\nOptimizeSpeed    = \"-O3\",\nSymbols          = \"-g\",\n}\nlocal cxxflags =\n{\nCpp11        = \"-std=c++11\",\nCpp14        = \"-std=c++14\",\nCpp17        = \"-std=c++17\",\nCppLatest    = \"-std=c++2a\",\nNoExceptions = \"-fno-exceptions\",\nNoRTTI       = \"-fno-rtti\",\nUnsignedChar = \"-funsigned-char\",\n}\nlocal objcflags =\n{\nObjcARC = \"-fobjc-arc\",\n}\npremake.gcc.platforms =\n{\nNative = {\n"
+-	"cppflags = \"-MMD -MP\",\n},\nx32 = {\ncppflags = \"-MMD -MP\",\nflags    = \"-m32\",\n},\nx64 = {\ncppflags = \"-MMD -MP\",\nflags    = \"-m64\",\n},\nUniversal = {\nar       = \"libtool\",\ncppflags = \"-MMD -MP\",\nflags    = \"-arch i386 -arch x86_64 -arch ppc -arch ppc64\",\n},\nUniversal32 = {\nar       = \"libtool\",\ncppflags = \"-MMD -MP\",\nflags    = \"-arch i386 -arch ppc\",\n},\nUniversal64 = {\nar       = \"libtool\",\ncppflags = \"-MMD -MP\",\nflags    = \"-arch x86_64 -arch ppc64\",\n},\nPS3 = {\ncc         = \"ppu-lv2-g++\",\ncxx        = \"ppu-lv2-g++\",\nar         = \"ppu-lv2-ar\",\ncppflags   = \"-MMD -MP\",\n},\nWiiDev = {\ncppflags    = \"-MMD -MP -I$(LIBOGC_INC) $(MACHDEP)\",\nldflags= \"-L$(LIBOGC_LIB) $(MACHDEP)\",\ncfgsettings = [[\n  ifeq ($(strip $(DEVKITPPC)),)\n    $(error \"DEVKITPPC environment variable is not set\")'\n  endif\n  include $(DEVKITPPC)/wii_rules']],\n},\nOrbis = {\ncc         = \"orbis-clang\",\ncxx        = \"orbis-clang++\",\nar         = \"orbis-ar\",\ncppflag"
++	"cppflags = \"-MMD -MP\",\n},\nx32 = {\ncppflags = \"-MMD -MP\"\n},\nx64 = {\ncppflags = \"-MMD -MP\",\nflags    = \"-m64\",\n},\nUniversal = {\nar       = \"libtool\",\ncppflags = \"-MMD -MP\",\nflags    = \"-arch i386 -arch x86_64 -arch ppc -arch ppc64\",\n},\nUniversal32 = {\nar       = \"libtool\",\ncppflags = \"-MMD -MP\",\nflags    = \"-arch i386 -arch ppc\",\n},\nUniversal64 = {\nar       = \"libtool\",\ncppflags = \"-MMD -MP\",\nflags    = \"-arch x86_64 -arch ppc64\",\n},\nPS3 = {\ncc         = \"ppu-lv2-g++\",\ncxx        = \"ppu-lv2-g++\",\nar         = \"ppu-lv2-ar\",\ncppflags   = \"-MMD -MP\",\n},\nWiiDev = {\ncppflags    = \"-MMD -MP -I$(LIBOGC_INC) $(MACHDEP)\",\nldflags= \"-L$(LIBOGC_LIB) $(MACHDEP)\",\ncfgsettings = [[\n  ifeq ($(strip $(DEVKITPPC)),)\n    $(error \"DEVKITPPC environment variable is not set\")'\n  endif\n  include $(DEVKITPPC)/wii_rules']],\n},\nOrbis = {\ncc         = \"orbis-clang\",\ncxx        = \"orbis-clang++\",\nar         = \"orbis-ar\",\ncppflag"
+ 	"s   = \"-MMD -MP\",\n},\nEmscripten = {\ncc         = \"$(EMSCRIPTEN)/emcc\",\ncxx        = \"$(EMSCRIPTEN)/em++\",\nar         = \"$(EMSCRIPTEN)/emar\",\ncppflags   = \"-MMD -MP\",\n}\n}\nlocal platforms = premake.gcc.platforms\nfunction premake.gcc.getcppflags(cfg)\nlocal flags = { }\ntable.insert(flags, platforms[cfg.platform].cppflags)\nif flags[1]:startswith(\"-MMD\") then\ntable.insert(flags, \"-MP\")\nend\nreturn flags\nend\nfunction premake.gcc.getcflags(cfg)\nlocal result = table.translate(cfg.flags, cflags)\ntable.insert(result, platforms[cfg.platform].flags)\nif cfg.system ~= \"windows\" and cfg.kind == \"SharedLib\" then\ntable.insert(result, \"-fPIC\")\nend\nreturn result\nend\nfunction premake.gcc.getcxxflags(cfg)\nlocal result = table.translate(cfg.flags, cxxflags)\nreturn result\nend\nfunction premake.gcc.getobjcflags(cfg)\nreturn table.translate(cfg.flags, objcflags)\nend\nfunction premake.gcc.getldflags(cfg)\nlocal result = { }\nif not cfg.flags.Symbols then\nif cfg.system == \"macosx\" then"
+ 	"\nelse\ntable.insert(result, \"-s\")\nend\nend\nif cfg.kind == \"Bundle\" then\ntable.insert(result, \"-bundle\")\nend\nif cfg.kind == \"SharedLib\" then\nif cfg.system == \"macosx\" then\ntable.insert(result, \"-dynamiclib\")\nelse\ntable.insert(result, \"-shared\")\nend\nif cfg.system == \"windows\" and not cfg.flags.NoImportLib then\ntable.insert(result, '-Wl,--out-implib=\"' .. cfg.linktarget.fullpath .. '\"')\nend\nend\nif cfg.kind == \"WindowedApp\" and cfg.system == \"windows\" then\ntable.insert(result, \"-mwindows\")\nend\nlocal platform = platforms[cfg.platform]\ntable.insert(result, platform.flags)\ntable.insert(result, platform.ldflags)\nreturn result\nend\nfunction premake.gcc.getlibdirflags(cfg)\nlocal result = { }\nfor _, value in ipairs(premake.getlinks(cfg, \"all\", \"directory\")) do\ntable.insert(result, '-L\\\"' .. value .. '\\\"')\nend\nreturn result\nend\nfunction premake.gcc.islibfile(p)\nif path.getextension(p) == \".a\" then\nreturn true\nend\nreturn false\nend\nfunction premake.gcc.ge"
+ 	"tlibfiles(cfg)\nlocal result = {}\nfor _, value in ipairs(premake.getlinks(cfg, \"system\", \"fullpath\")) do\nif premake.gcc.islibfile(value) then\ntable.insert(result, _MAKE.esc(value))\nend\nend\nreturn result\nend\nfunction premake.gcc.getlinkflags(cfg)\nlocal result = {}\nfor _, value in ipairs(premake.getlinks(cfg, \"system\", \"fullpath\")) do\nif premake.gcc.islibfile(value) then\nvalue = path.rebase(value, cfg.project.location, cfg.location)\ntable.insert(result, _MAKE.esc(value))\nelseif path.getextension(value) == \".framework\" then\ntable.insert(result, '-framework ' .. _MAKE.esc(path.getbasename(value)))\nelse\ntable.insert(result, '-l' .. _MAKE.esc(path.getname(value)))\nend\nend\nreturn result\nend\nfunction premake.gcc.wholearchive(lib)\nif premake.gcc.llvm then\nreturn {\"-force_load\", lib}\nelse\nreturn {\"-Wl,--whole-archive\", lib, \"-Wl,--no-whole-archive\"}\nend\nend\nfunction premake.gcc.getarchiveflags(prj, cfg, ndx)\nlocal result = {}\nif cfg.platform:startswith(\"Universal\") then\n"
+--- a/scripts/toolchain.lua	2019-04-30 15:47:21.000000000 +0200
++++ b/scripts/toolchain.lua	2019-05-21 22:51:48.116148725 +0200
+@@ -698,9 +698,6 @@
+ 
+ 	configuration { "linux-gcc", "x32" }
+ 		objdir (_buildDir .. "linux_gcc" .. "/obj")
+-		buildoptions {
+-			"-m32",
+-		}
+ 
+ 	configuration { "linux-gcc", "x32", "Release" }
+ 		targetdir (_buildDir .. "linux_gcc" .. "/bin/x32/Release")

--- a/packages/lakka/libretro_cores/mame/patches/mame-remove-bgfx.patch
+++ b/packages/lakka/libretro_cores/mame/patches/mame-remove-bgfx.patch
@@ -1,0 +1,80 @@
+diff --git a/Makefile.libretro b/Makefile.libretro
+index 31c414be..8de62064 100644
+--- a/Makefile.libretro
++++ b/Makefile.libretro
+@@ -55,6 +55,8 @@ VERBOSE ?= 1
+ # scripts/toolchain.lua)
+ # PTR64 = 1
+ 
++USE_BGFX ?= 0
++
+ ###########################################################################
+ #
+ #   LIBRETRO PLATFORM GUESSING
+diff --git a/scripts/genie.lua b/scripts/genie.lua
+index 4655b3f4..495308a8 100644
+--- a/scripts/genie.lua
++++ b/scripts/genie.lua
+@@ -387,8 +387,22 @@ newoption {
+ 	description = "Arguments for running debug build.",
+ }
+ 
++newoption {
++	trigger = "USE_BGFX",
++	description = "Use bgfx.",
++	allowed = {
++		{ "0",   "Disabled"     },
++		{ "1",   "Enabled"      },
++	}
++}
++
+ dofile ("extlib.lua")
+ 
++if not _OPTIONS["USE_BGFX"] then
++	_OPTIONS["USE_BGFX"] = "0"
++end
++
++
+ if _OPTIONS["SHLIB"]=="1" then
+ 	LIBTYPE = "SharedLib"
+ else
+diff --git a/scripts/src/3rdparty.lua b/scripts/src/3rdparty.lua
+index 664898d1..d720578d 100644
+--- a/scripts/src/3rdparty.lua
++++ b/scripts/src/3rdparty.lua
+@@ -1249,7 +1249,7 @@
+ --------------------------------------------------
+ -- BGFX library objects
+ --------------------------------------------------
+-
++if not _OPTIONS["use_bgfx"]=="0" then
+ project "bgfx"
+ 	uuid "d3e7e119-35cf-4f4f-aba0-d3bdcd1b879a"
+ 	kind "StaticLib"
+@@ -1433,7 +1433,7 @@
+ 			"-x objective-c++",
+ 		}
+ 	end
+-
++end
+ 
+ --------------------------------------------------
+ -- PortAudio library objects
+diff --git a/scripts/src/main.lua b/scripts/src/main.lua
+index 0b47e662..b19eb646 100644
+--- a/scripts/src/main.lua	2019-05-22 00:15:58.983419982 +0200
++++ b/scripts/src/main.lua	2019-05-22 00:17:39.675345144 +0200
+@@ -380,8 +380,12 @@
+ 			ext_lib("portmidi"),
+ 		}
+ 	end
++	if _OPTIONS["USE_BGFX"]~="0" then
++		links {
++			"bgfx",
++		}
++	end
+ 	links {
+-		"bgfx",
+ 		"bimg",
+ 		"bx",
+ 		"ocore_" .. _OPTIONS["osd"],

